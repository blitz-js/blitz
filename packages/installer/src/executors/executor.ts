import {log} from '@blitzjs/display'

export interface BaseExecutor {
  stepId: string | number
  stepName: string
  // a bit to display to the user to give context to the change
  explanation: string
}

type dynamicExecutorArgument<T> = (cliArgs: any) => T

function isDynamicExecutorArgument<T>(input: executorArgument<T>): input is dynamicExecutorArgument<T> {
  return typeof (input as dynamicExecutorArgument<T>) === 'function'
}

export type executorArgument<T> = T | dynamicExecutorArgument<T>

export function logExecutorFrontmatter(executor: BaseExecutor) {
  console.log()
  const lineLength = executor.stepName.length + 6
  const verticalBorder = `+${new Array(lineLength).fill('–').join('')}+`
  log.branded(verticalBorder)
  log.branded(`⎪   ${executor.stepName}   ⎪`)
  log.branded(verticalBorder)
  log.info(executor.explanation)
  console.log()
}

export function getExecutorArgument<T>(input: executorArgument<T>, cliArgs: any): T {
  if (isDynamicExecutorArgument(input)) {
    return input(cliArgs)
  }
  return input
}
