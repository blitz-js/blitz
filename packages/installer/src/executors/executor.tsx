import * as React from "react"
import {Box, Text} from "ink"
import {Newline} from "../components/newline"
import {Branded} from "../components/branded"

export interface ExecutorConfig {
  stepId: string | number
  stepName: string
  stepType: string
  // a bit to display to the user to give context to the change
  explanation: string
}

export interface Executor {
  type: string
  Propose: React.FC<{step: ExecutorConfig; onProposalAccepted: (data?: any) => void; cliArgs: any}>
  Commit: React.FC<{
    step: ExecutorConfig
    proposalData?: any
    onChangeCommitted: (data?: any) => void
    cliArgs: any
  }>
}

type dynamicExecutorArgument<T> = (cliArgs: any) => T

function isDynamicExecutorArgument<T>(
  input: executorArgument<T>,
): input is dynamicExecutorArgument<T> {
  return typeof (input as dynamicExecutorArgument<T>) === "function"
}

export type executorArgument<T> = T | dynamicExecutorArgument<T>

export function Frontmatter({executor}: {executor: ExecutorConfig}) {
  const lineLength = executor.stepName.length + 6
  const verticalBorder = `+${new Array(lineLength).fill("–").join("")}+`
  return (
    <Box flexDirection="column" paddingBottom={1}>
      <Newline />
      <Branded>
        <Box flexDirection="column">
          <Text>{verticalBorder}</Text>
          <Text>⎪&nbsp;&nbsp;&nbsp;{executor.stepName}&nbsp;&nbsp;&nbsp;⎪</Text>
          <Text>{verticalBorder}</Text>
        </Box>
      </Branded>
      <Text bold>{executor.explanation}</Text>
    </Box>
  )
}

export function getExecutorArgument<T>(input: executorArgument<T>, cliArgs: any): T {
  if (isDynamicExecutorArgument(input)) {
    return input(cliArgs)
  }
  return input
}
