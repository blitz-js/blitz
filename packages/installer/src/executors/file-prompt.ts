import {prompt as enquirer} from "enquirer"
import globby from "globby"

enum SearchType {
  file,
  directory,
}

interface FilePromptOptions {
  globFilter?: string
  getChoices?(context: any): string[]
  searchType?: SearchType
  context: any
}

function getMatchingFiles(filter: string = ""): Promise<string[]> {
  return globby(filter, {expandDirectories: true})
}

export async function filePrompt(options: FilePromptOptions): Promise<string> {
  const choices = options.getChoices
    ? options.getChoices(options.context)
    : await getMatchingFiles(options.globFilter)
  if (choices.length === 1) {
    return choices[0]
  }
  const results: {file: string} = await enquirer({
    type: "autocomplete",
    name: "file",
    message: "Select the target file",
    // @ts-ignore
    limit: 10,
    choices,
  })
  return results.file
}
