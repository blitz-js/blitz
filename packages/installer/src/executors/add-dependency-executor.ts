import {BaseExecutor, executorArgument, getExecutorArgument} from './executor'
import * as fs from 'fs-extra'
import * as path from 'path'
import spawn from 'cross-spawn'
import {log} from '@blitzjs/display'

interface NpmPackage {
  name: string
  // defaults to latest published
  version?: string
  // defaults to false
  isDevDep?: boolean
}

export interface AddDependencyExecutor extends BaseExecutor {
  packages: executorArgument<NpmPackage[]>
}

export function isAddDependencyExecutor(executor: BaseExecutor): executor is AddDependencyExecutor {
  return (executor as AddDependencyExecutor).packages !== undefined
}

async function getPackageManager(): Promise<'yarn' | 'npm'> {
  if (fs.existsSync(path.resolve('package-lock.json'))) {
    return 'npm'
  }
  return 'yarn'
}

export async function addDependencyExecutor(executor: AddDependencyExecutor, cliArgs: any): Promise<void> {
  const packageManager = await getPackageManager()
  const packagesToInstall = getExecutorArgument(executor.packages, cliArgs)
  for (const pkg of packagesToInstall) {
    const args: string[] = ['add']
    // if devDep flag isn't specified we install as a regular dependency, so
    // we need to explicitly check for `true`
    if (pkg.isDevDep === true) {
      args.push(packageManager === 'yarn' ? '-D' : '--save-dev')
    }
    pkg.version ? args.push(`${pkg.name}@${pkg.version}`) : args.push(pkg.name)
    log.meta(`Installing ${pkg.name} ${pkg.isDevDep !== false ? 'as a dev dependency' : ''}`)
    spawn.sync(packageManager, args, {
      stdio: ['inherit', 'pipe', 'pipe'],
    })
  }
  log.progress(`${packagesToInstall.length} packages installed successfully`)
}
