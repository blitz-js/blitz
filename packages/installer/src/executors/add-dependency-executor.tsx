import {ExecutorConfig, executorArgument, getExecutorArgument, Executor} from "./executor"
import * as fs from "fs-extra"
import * as path from "path"
import {spawn} from "cross-spawn"
import * as React from "react"
import {Box, Text} from "ink"
import {Newline} from "../components/newline"
import Spinner from "ink-spinner"
import {useEnterToContinue} from "../utils/use-enter-to-continue"

interface NpmPackage {
  name: string
  // defaults to latest published
  version?: string
  // defaults to false
  isDevDep?: boolean
}

export interface Config extends ExecutorConfig {
  packages: executorArgument<NpmPackage[]>
}

export function isAddDependencyExecutor(executor: ExecutorConfig): executor is Config {
  return (executor as Config).packages !== undefined
}

export const type = "add-dependency"

function Package({pkg, loading}: {pkg: NpmPackage; loading: boolean}) {
  return (
    <Text>
      {`   `}
      {loading ? <Spinner /> : "ðŸ“¦"}
      {` ${pkg.name}@${pkg.version}`}
    </Text>
  )
}

const DependencyList = ({
  lede,
  depsLoading = false,
  devDepsLoading = false,
  packages,
}: {
  lede: string
  depsLoading?: boolean
  devDepsLoading?: boolean
  packages: NpmPackage[]
}) => {
  return (
    <Box flexDirection="column">
      <Text>{lede}</Text>
      <Newline />
      <Text>Dependencies to be installed:</Text>
      {packages
        .filter((p) => !p.isDevDep)
        .map((pkg) => (
          <Package key={pkg.name} pkg={pkg} loading={depsLoading} />
        ))}
      <Newline />
      <Text>Dev Dependencies to be installed:</Text>
      {packages
        .filter((p) => p.isDevDep)
        .map((pkg) => (
          <Package key={pkg.name} pkg={pkg} loading={devDepsLoading} />
        ))}
    </Box>
  )
}

export const Propose: Executor["Propose"] = ({cliArgs, step, onProposalAccepted}) => {
  useEnterToContinue(onProposalAccepted)

  if (!isAddDependencyExecutor(step)) {
    onProposalAccepted()
    return null
  }
  return (
    <DependencyList
      lede={"Adding some shiny new dependencies.\nIf this list looks good, press ENTER to install."}
      packages={getExecutorArgument(step.packages, cliArgs)}
    />
  )
}

function getPackageManager() {
  if (fs.existsSync(path.resolve("package-lock.json"))) {
    return "npm"
  }
  return "yarn"
}

async function installPackages(packages: NpmPackage[], isDev = false) {
  const packageManager = getPackageManager()
  const args: string[] = ["add"]

  if (isDev) {
    args.push(packageManager === "yarn" ? "-D" : "--save-dev")
  }
  packages.forEach((pkg) => {
    pkg.version ? args.push(`${pkg.name}@${pkg.version}`) : args.push(pkg.name)
  })
  await new Promise((resolve) => {
    const cp = spawn(packageManager, args, {
      stdio: ["inherit", "pipe", "pipe"],
    })
    cp.on("exit", resolve)
  })
}

export const Commit: Executor["Commit"] = ({cliArgs, step, onChangeCommitted}) => {
  const [depsInstalled, setDepsInstalled] = React.useState(false)
  const [devDepsInstalled, setDevDepsInstalled] = React.useState(false)

  const handleChangeCommitted = React.useCallback(() => {
    const packages = (step as Config).packages
    const dependencies = packages.length === 1 ? "dependency" : "dependencies"
    onChangeCommitted(`Installed ${packages.length} ${dependencies}`)
  }, [onChangeCommitted, step])

  useEnterToContinue(handleChangeCommitted, depsInstalled && devDepsInstalled)

  React.useEffect(() => {
    async function installDeps() {
      const packagesToInstall = getExecutorArgument((step as Config).packages, cliArgs).filter(
        (p) => !p.isDevDep,
      )
      await installPackages(packagesToInstall)
      setDepsInstalled(true)
    }
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    installDeps()
  }, [cliArgs, step])

  React.useEffect(() => {
    if (!depsInstalled) return
    async function installDevDeps() {
      const packagesToInstall = getExecutorArgument((step as Config).packages, cliArgs).filter(
        (p) => p.isDevDep,
      )
      await installPackages(packagesToInstall, true)
      setDevDepsInstalled(true)
    }
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    installDevDeps()
  }, [cliArgs, depsInstalled, step])

  if (!isAddDependencyExecutor(step)) {
    onChangeCommitted()
    return null
  }
  return (
    <>
      <DependencyList
        lede={"Hang tight! Fetching the latest dependencies..."}
        depsLoading={!depsInstalled}
        devDepsLoading={!devDepsInstalled}
        packages={getExecutorArgument(step.packages, cliArgs)}
      />
      {depsInstalled && devDepsInstalled && (
        <Box paddingTop={1}>
          <Text>Dependencies installed! Press ENTER to continue</Text>
        </Box>
      )}
    </>
  )
}
