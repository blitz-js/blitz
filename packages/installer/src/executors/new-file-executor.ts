import {BaseExecutor, executorArgument, getExecutorArgument} from './executor'
import {Generator, GeneratorOptions} from '@blitzjs/generator'
import {log} from '@blitzjs/display'
import {waitForConfirmation} from '../utils/wait-for-confirmation'

export interface NewFileExecutor extends BaseExecutor {
  targetDirectory?: executorArgument<string>
  templatePath: executorArgument<string>
  templateValues: executorArgument<{[key: string]: string}>
  destinationPathPrompt?: executorArgument<string>
}

export function isNewFileExecutor(executor: BaseExecutor): executor is NewFileExecutor {
  return (executor as NewFileExecutor).templatePath !== undefined
}

interface TempGeneratorOptions extends GeneratorOptions {
  targetDirectory?: string
  templateRoot: string
  templateValues: any
}

class TempGenerator extends Generator<TempGeneratorOptions> {
  sourceRoot: string
  targetDirectory: string
  templateValues: any

  constructor(options: TempGeneratorOptions) {
    super(options)
    this.sourceRoot = options.templateRoot
    this.templateValues = options.templateValues
    this.targetDirectory = options.targetDirectory || '.'
  }

  getTemplateValues() {
    return this.templateValues
  }

  getTargetDirectory() {
    return this.targetDirectory
  }
}

export async function newFileExecutor(executor: NewFileExecutor, cliArgs: any): Promise<void> {
  const generatorArgs = {
    destinationRoot: '.',
    targetDirectory: getExecutorArgument(executor.targetDirectory, cliArgs),
    templateRoot: getExecutorArgument(executor.templatePath, cliArgs),
    templateValues: getExecutorArgument(executor.templateValues, cliArgs),
  }
  const dryRunGenerator = new TempGenerator({
    ...generatorArgs,
    dryRun: true,
  })
  const commitGenerator = new TempGenerator(generatorArgs)
  log.progress("First we'll do a dry-run. Here's a list of files that would be created:")
  await dryRunGenerator.run()
  await waitForConfirmation('To commit the changes, press enter. Press Ctrl+C to abort')
  await commitGenerator.run()
}
