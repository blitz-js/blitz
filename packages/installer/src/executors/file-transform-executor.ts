import {BaseExecutor, executorArgument, getExecutorArgument} from './executor'
import {filePrompt} from './file-prompt'
import {transform, Transformer} from '../utils/transform'
import {log} from '@blitzjs/display'
import {waitForConfirmation} from '../utils/wait-for-confirmation'
import {createPatch} from 'diff'
import chokidar from 'chokidar'
import * as fs from 'fs-extra'
import chalk from 'chalk'

export interface FileTransformExecutor extends BaseExecutor {
  selectTargetFiles?(cliArgs: any): any[]
  singleFileSearch?: executorArgument<string>
  transform: Transformer
}

export function isFileTransformExecutor(executor: BaseExecutor): executor is FileTransformExecutor {
  return (executor as FileTransformExecutor).transform !== undefined
}

async function executeWithDiff(transformFn: Transformer, filePath: string) {
  await new Promise((res, rej) => {
    const watcher = chokidar.watch(filePath)
    const originalFileContents = fs.readFileSync(filePath).toString('utf-8')
    watcher.on('change', (path) => {
      watcher.close().then(() => {
        const patch = createPatch(path, originalFileContents, fs.readFileSync(path).toString('utf-8'))
        patch
          .split('\n')
          .slice(2)
          .forEach((line) => {
            if (line.startsWith('-') && !line.startsWith('---')) console.log(chalk.bold.red(line))
            else if (line.startsWith('+') && !line.startsWith('+++')) console.log(chalk.bold.green(line))
            else console.log(line)
          })
        res(path)
      })
    })
    watcher.on('error', (error) => {
      rej(error)
    })
    transform(transformFn, [filePath])
  })
}

export async function fileTransformExecutor(executor: FileTransformExecutor, cliArgs: any): Promise<void> {
  const fileToTransform: string = await filePrompt({
    context: cliArgs,
    globFilter: getExecutorArgument(executor.singleFileSearch, cliArgs),
    getChoices: executor.selectTargetFiles,
  })
  try {
    await executeWithDiff(executor.transform, fileToTransform)
    await waitForConfirmation('The above changes were applied. Press enter to continue')
  } catch (err) {
    log.error(`Failed to transform ${fileToTransform}`)
  }
}
