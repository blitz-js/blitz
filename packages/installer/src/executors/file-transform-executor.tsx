import {ExecutorConfig, executorArgument, getExecutorArgument, Executor} from "./executor"
import {filePrompt} from "./file-prompt"
import {processFile, transform, Transformer, TransformStatus} from "../utils/transform"
import {createPatch} from "diff"
import * as fs from "fs-extra"
import * as React from "react"
import Spinner from "ink-spinner"
import {Box, Text, Color} from "ink"
import {Newline} from "../components/newline"
import {useEnterToContinue} from "../utils/use-enter-to-continue"

export interface Config extends ExecutorConfig {
  selectTargetFiles?(cliArgs: any): any[]
  singleFileSearch?: executorArgument<string>
  transform: Transformer
}

export function isFileTransformExecutor(executor: ExecutorConfig): executor is Config {
  return (executor as Config).transform !== undefined
}

export const type = "file-transform"
export const Propose: Executor["Propose"] = ({cliArgs, onProposalAccepted, step}) => {
  const [diff, setDiff] = React.useState<string | null>(null)
  const filePathRef = React.useRef("")
  useEnterToContinue(() => onProposalAccepted(filePathRef.current), !!filePathRef.current)

  React.useEffect(() => {
    async function generateDiff() {
      const fileToTransform: string = await filePrompt({
        context: cliArgs,
        globFilter: getExecutorArgument((step as Config).singleFileSearch, cliArgs),
        getChoices: (step as Config).selectTargetFiles,
      })
      filePathRef.current = fileToTransform
      const originalFile = fs.readFileSync(fileToTransform).toString("utf-8")
      const newFile = processFile(originalFile, (step as Config).transform)
      return createPatch(fileToTransform, originalFile, newFile)
    }
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    generateDiff().then(setDiff)
  }, [cliArgs, step])

  if (!diff) {
    return (
      <Box>
        <Spinner /> Generating file diff...
      </Box>
    )
  }

  return (
    <Box flexDirection="column">
      {diff
        .split("\n")
        .slice(2)
        .map((line, idx) => {
          let styleProps: any = {}
          let colorProps: any = {}
          if (line.startsWith("-") && !line.startsWith("---")) {
            styleProps.bold = true
            colorProps.red = true
          } else if (line.startsWith("+") && !line.startsWith("+++")) {
            styleProps.bold = true
            colorProps.green = true
          }

          return (
            <Color {...colorProps} key={idx}>
              <Text {...styleProps}>{line}</Text>
            </Color>
          )
        })}
      <Newline />
      <Text>
        The above diff will be applied. If it looks okay to you, hit ENTER to apply the changes!
      </Text>
    </Box>
  )
}
export const Commit: Executor["Commit"] = ({onChangeCommitted, proposalData: filePath, step}) => {
  const transformFn = (step as Config).transform

  const [loading, setLoading] = React.useState(true)

  const handleChangeCommitted = React.useCallback(() => {
    onChangeCommitted(`Modified 1 file: ${filePath}`)
  }, [filePath, onChangeCommitted])

  useEnterToContinue(handleChangeCommitted, !loading)

  React.useEffect(() => {
    const results = transform(transformFn, [filePath])
    if (results.some((r) => r.status === TransformStatus.Failure)) {
      console.error(results)
    }
    setLoading(false)
  }, [filePath, transformFn])

  if (loading) {
    return (
      <Box>
        <Spinner /> Applying file changes
      </Box>
    )
  }
  return (
    <Box paddingBottom={1} flexDirection="column">
      {/* Stop ESlint from complaining about the emoji */}
      {/* eslint-disable-next-line */}
      <Text>üìù File changes applied! Press ENTER to continue</Text>
    </Box>
  )
}
