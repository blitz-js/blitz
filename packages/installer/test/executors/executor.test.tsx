import {render} from "ink-testing-library"
import React from "react"
import stripAnsi from "strip-ansi"
import {Frontmatter} from "../../src/executors/executor"

describe("Executor", () => {
  const executorConfig = {
    stepId: "newFile",
    stepName: "New File",
    stepType: "new-file",
    explanation: "Testing text for a new file",
  }
  it("should render Frontmatter", () => {
    const {lastFrame} = render(<Frontmatter executor={executorConfig} />)

    expect(stripAnsi(lastFrame())).toMatchSnapshot()
  })

  it("should contain a step name and explanation", () => {
    const {frames} = render(<Frontmatter executor={executorConfig} />)

    expect(frames[0].includes("New File")).toBeTruthy()
    expect(frames[0].includes("Testing text for a new file")).toBeTruthy()
  })
})
