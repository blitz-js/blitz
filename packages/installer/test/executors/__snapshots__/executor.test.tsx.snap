// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Executor should render Frontmatter 1`] = `
"
+––––––––––––––+
⎪   New File   ⎪
+––––––––––––––+
Testing text for a new file
"
`;
